File: python/NLSolver.py
import numpy as np
from numpy.linalg import pinv

class NLSolver():
    '''
    Class which can implements Gauss-Newton and Levenberg–Marquardt non-linear regression

    Taken directly from here:
    https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm

    Note, I found the wikipedia article about Gauss-Newton less clear than this article,
    and entire assumption of this project is these two are quite similar, and can be 
    encapulsated in a single class with shared methods.

    Constructor:
    b_init: np.array - initial guess for starting parameters
    x: np.array - x values for which you have reference data
    y: np.array - y references values, used as target for fit
    func: function - fitting function, must have form func(x,b), where b is np.array of len(b_init)
    method: str - Gauss-Newton and Levenberg–Marquardt
    lambd: float - Levenberg–Marquardt includes a damping parameter, which needs an initial value
    '''
    
    def __init__(self,
                 b_init:np.array, 
                 x:np.array,
                 y:np.array,
                 func:any,
                 method:str='Gauss-Newton',
                 lambd_init: float=1e-2):

        self.b = b_init
        self.x = x
        self.y = y
        self.func = func
        self.y_pred = self.func(x,self.b)
        self.lambd = lambd_init

        assert method in ['Gauss-Newton','Levenberg–Marquardt']
        self.method = method

    def fit(self,
            step_init:float=1e-6, 
            max_iter:int=100,
            stop_tolerance:float=1e-10):
        '''
        Canonical fit method, which will run through non-linear fit, and update self.b (parameters of fit)

        Input:
        step_init:float - initial step size delta for first round of iteration
        max_iter: int - maximum number of iterations of fitting
        stop_tolerance: float - if absolute value of rmse - rmse_prev is < this tolerance, considered converged
        '''

        b0 = self.b.copy()
        self.b = self.b + step_init

        if self.method == 'Gauss-Newton':
            for i in range(max_iter):
                rmse_prev = self.compute_rmse(self.y, self.func, self.x, self.b)
                r = self.compute_residual(self.y, self.func, self.x, self.b)
                J = self.compute_jacobian(self.x, self.b, b0, self.func, self.y)
                JpI = self.compute_pseudoinverse(J)
                b0 = self.b.copy()
                self.b = self.b + JpI @ r
                rmse = self.compute_rmse(self.y, self.func, self.x, self.b)

                print('Gauss-Newton; Iteration: %s, RMSE: %s'%(i,rmse))
                if np.abs(rmse_prev - rmse) < stop_tolerance:
                    break
                elif np.isnan(self.b).sum() > 0:
                    print('Fit diverged, trying different starting values')
                    break

        elif self.method == 'Levenberg–Marquardt':
            for i in range(max_iter):
                rmse_prev = self.compute_rmse(self.y, self.func, self.x, self.b)
                r = self.compute_residual(self.y, self.func, self.x, self.b)
                J = self.compute_jacobian(self.x, self.b, b0, self.func, self.y)
                JpI = self.compute_lm_inverse(J, self.lambd)
                b0 = self.b.copy()
                self.b = self.b + JpI @ r
                rmse = self.compute_rmse(self.y, self.func, self.x, self.b)

                print('Levenberg–Marquardt; Iteration: %s, RMSE: %s'%(i,rmse))
                if np.abs(rmse_prev - rmse) < stop_tolerance:
                    break
                elif np.isnan(self.b).sum() > 0:
                    print('Fit diverged, trying different starting values')
                    break
                # This is a very simple approach to updating lambda, 
                # called delayed gratification according to wikipedia
                elif rmse < rmse_prev:
                    self.lambd *= 0.33
                elif rmse > rmse_prev:
                    self.lambd *= 2


    def compute_jacobian(self, x:np.array, b:np.array, 
                         b0:np.array, func:any, y_ref:np.array):
        '''
        Numerically compute Jacobian with respect to b, at points x

        I found this very helpful when writing this:
        https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives/jacobian/

        Input:
        x: np.array - points at which to evluate jacobian
        b: np.array - parameters to take jacobian wrt
        b0: np.array - previous parameters, informs size of local region for linearized approximation
        func: function - used to evaluate y at each point, in order to calculate residual
        y_ref: np.array - ground truth values, for computing residuals 

        Return:
        J: np.array - Jacobian wrt b, in local region b0-b at points x
        '''
        b = b.astype(float)
        b0 = b0.astype(float)
        J = np.zeros((len(x),len(b)))
        y0 = self.compute_residual(y_ref, func, x, b0)
        for i in range(J.shape[1]):
            b_partial = b0.copy()
            b_partial[i] = b[i]
            delta_residual = y0 - self.compute_residual(y_ref, func, x, b_partial)
            J[:,i] = delta_residual / (b[i] - b0[i])
        return J

    def predict(self):
        self.y_pred = self.func(self.x, self.b)
        return self.y_pred

    @staticmethod
    def compute_residual(y_ref:np.array, func:any, x:np.array, b:np.array):
        '''
        Compute residual based on reference, function, points and parameters

        Input:
        y_ref: np.array - ground truth values, for computing residuals 
        func: function - used to evaluate y at each point
        x: np.array - points at which to evluate residuals
        b: np.array - parameters for func
        '''
        return y_ref - func(x, b)

    @staticmethod
    def compute_rmse(y_ref:np.array, func:any, x:np.array, b:np.array):
        '''
        Compute rmse based on reference, function, points and parameters

        Input:
        y_ref: np.array - ground truth values, for computing residuals 
        func: function - used to evaluate y at each point
        x: np.array - points at which to evluate residuals
        b: np.array - parameters for func
        '''
        r = y_ref - func(x, b)
        return np.sqrt((r**2).mean())

    @staticmethod
    def compute_pseudoinverse(X:np.array):
        '''
        Compute Moore-Penrose pseuodinverse, other methods possible here (i.e., np.linalg.solve)

        Inputs:
        X: np.array - Input matrix to invert
        '''
        return pinv(X.T @ X) @ X.T

    @staticmethod
    def compute_lm_inverse(X, lambd):
        '''
        Compute inverse matrix for Levenberg–Marquardt, same as Gauss-Newton with added damping term

        Inputs:
        X: np.array - Input matrix to invert
        lambd: float - Damping factor for LM method
        '''
        return pinv(X.T @ X + lambd*np.diag(X.T @ X)) @ X.T

File: cpp/NLSolver/NLSolver.cpp
#include "NLSolver.h"
#include <stdexcept>
#include <cmath>

using namespace std;

NLSolver::NLSolver() {}

NLSolver::NLSolver(VectorXd b_init, VectorXd x, VectorXd y, 
                   function<VectorXd(VectorXd, VectorXd)> func,
                   string method, float lambd_init) {
    
    b = b_init;
    y_pred = this->func(x, b_init);
    this->x = x;
    this->y = y;
    cout << "Constructed okay" << endl;
    this->func = func;

    if (method == "Gauss-Newton") {
        this->method = method;
        lambd = 0;
    } else if (method == "Levenberg–Marquardt") {
        this->method = method;
    } else {
        throw std::invalid_argument("Invalid ");
    }
        
}

MatrixXd NLSolver::compute_jacobian(VectorXd b, VectorXd b0) {

    int lenx = x.cols();
    int lenb = b.cols();
    VectorXd b_part(lenb);

    MatrixXd J(lenx, lenb);
    VectorXd y0 = compute_residual(b);
    VectorXd delta_res(y0.cols());

    for (int i = 0; i < lenb; i++) {
        b_part = b0;
        b_part(i) = b(i);
        delta_res = y0 - compute_residual(b_part);
        J(all,i) = delta_res / (b(i) - b0(i));
    }
    return J;        
}

VectorXd NLSolver::predict() {
    y_pred = func(x, b);
    return y_pred; 
}

VectorXd NLSolver::compute_residual(VectorXd b) {
    VectorXd y_pred = func(x, b);
    return y - y_pred; 
}

float NLSolver::compute_rmse(VectorXd b) {
    VectorXd r = compute_residual(b);
    return sqrt(r.array().pow(2).mean());
}

MatrixXd NLSolver::compute_pseudoinverse(MatrixXd X) {
    MatrixXd Xt = X.transpose();
    return ((Xt * X).inverse()) * Xt;
}

MatrixXd NLSolver::compute_lm_inverse(MatrixXd X, float lambd) {
    MatrixXd Xt = X.transpose();
    return ((Xt * X + lambd*(Xt * X).diagonal()).inverse()) * Xt;
}

VectorXd NLSolver::fit(float step_init, int max_iter, float stop_tolerance) {
    
    VectorXd b0 = b;
    float rmse_prev;
    b.array() += step_init;
    float rmse = compute_rmse(b);

    VectorXd r(y.cols());
    MatrixXd J(x.cols(), b.cols());
    MatrixXd Ji(b.cols(), x.cols());

    if (method == "Gauss-Newton") {

        for (int i = 0; i < max_iter; i++) {
            rmse_prev = rmse;
            r = compute_residual(b);
            J = compute_jacobian(b, b0);
            Ji = compute_pseudoinverse(J);
            b0 = b;
            b += Ji * r;
            rmse = compute_rmse(b);
            cout << "Gauss-Newton" << "Iteration: " << i << " RMSE: " << rmse << endl;

            if (abs(rmse_prev - rmse) < stop_tolerance)
                break;
        }
        return b;

    } else if (method == "Levenberg-Marquardt") {

        for (int i = 0; i < max_iter; i++) {
            rmse_prev = rmse;
            r = compute_residual(b);
            J = compute_jacobian(b, b0);
            Ji = compute_lm_inverse(J, lambd);
            b0 = b;
            b += Ji * r;
            rmse = compute_rmse(b);
            cout << "Levenberg-Marquardt" << "Iteration: " << i << " RMSE: " << rmse << endl;

            if (abs(rmse_prev - rmse) < stop_tolerance)
                break;
            else if (rmse < rmse_prev)
                lambd *= 0.33;
            else
                lambd *= 2;
        }
        return b;
    } else {
        for (int i = 0; i < b.cols(); i++) {
            b << -999;
        }
        cout << "Invalid Solver Type " << method << endl;
        return b;
    }

}

VectorXd rate_eq(VectorXd x, VectorXd args) {
    VectorXd r(x.cols());
    r = args(0)*x.array() / (args(1) + x.array());
    return r;
};

int main() {

    // Inputs, from wikipedia example
    VectorXd S(7), r(7), b0(2);
    S << 0.038,0.194,0.425,0.626,1.253,2.5,3.74;
    r << 0.05,0.127,0.094,0.2122,0.2729,0.2665,0.3317;
    b0 << 0.1,0.1;
    cout << "Well now that its all setup..." << endl;

    NLSolver nls = NLSolver(b0, S, r, &rate_eq, "Gauss-Newton", 1e-3);
    cout << "An instance of a thing..." << endl;
    VectorXd bfit = nls.fit(1e-6, 100, 1e-10);
    cout << "Finished Fitting, Final B Values: " << bfit << endl;

}